### ISSUE TITLE
Unfolding unapplied definition

### ISSUE THREAD

**Issue by Kha** [leanprover/lean3#1331]
The following proof fails if the `intro` is removed:
```lean
  notation a ≤ b ≤ c := a ≤ b ∧ b ≤ c

  -- TODO: `decidable_linear_order (fin n)`?
  def is_digit (c : char) : Prop :=
  #"0"^.to_nat ≤ c^.to_nat ≤ #"9"^.to_nat

  instance is_digit.decidable_pred : decidable_pred is_digit :=
  begin intro, dunfold is_digit, apply_instance end
```

Apparently `tactic_dunfold_expr_core` tries to apply `is_digit`'s equation lemma, which talks about the applied function. Is there even any benefit to that lemma? If I patch the implementation to just ignore eqn lemmas, the proof goes through.

**Comment by Kha** [comment]
Even for equational definitions, it may sometimes be nicer to have a "raw" unfold:
```lean
  def is_space : char → Prop
  | #" "    := true
  | #"\x09" := true -- \t
  | #"\n"   := true
  | #"\x0d" := true -- \r
  | _       := false

  -- HACK: instead of dunfold
  local attribute [reducible] is_space is_space._main
  instance is_space.decidable_pred : decidable_pred is_space :=
  by apply_instance
```

Edit: The planned auto-generation of definition induction principles would also solve this.

**Comment by leodemoura** [comment]
The goal is to make sure all definitions have a set of equational lemmas. This is useful for tactics such as `simp` and `ematch_using`. Both of them allow us to provide a set of lemmas. For convenience, we allow the user to write `simp [f]` as a short-hand for `simp [f.equations._eqn_1, ..., f.equations._eqn_n]`. Same for `ematch_using`. It would be awkward if this notation worked only for definitions defined using the equation compiler. Another motivation is to get uniform behavior at `dunfold`. It now always uses the equational lemmas. An alternative implementation would be: always apply delta reduction, but this is bad for functions defined using the equation compiler since it would potentially expose a bunch of nasty recursor applications. Here are two possibilities to address this issue.

- Add a "strong unfold" tactic (`sunfold`) that just applies delta reduction. 

- Use `dsimp [f]` or `simp [f]` for unfolding `f` using equational lemmas, and change `dunfold` implementation to be the strong unfold based on delta reduction.

What do you think?

**Comment by Kha** [comment]
I see, I did not know about the short-hand. What are the current differences between `dsimp [f]` and `dunfold [f]`? If they are not too great, I would be in favor of the second approach; it seems Coq's `unfold` also works that way.
Isabelle's `unfold` is also interesting: It takes arbitrary equations as input and tries to rewrite using them, falling back to trying them again _after applying function extensionality_ to also match under-applied occurences.

**Comment by leodemoura** [comment]
> What are the current differences between dsimp [f] and dunfold [f]?

`dsimp [f]` will also apply all refl lemmas tagged with `[simp]`, will expand reducible constants, and will perform beta reduction, and reduce projection over constructor.

> Isabelle's unfold is also interesting: It takes arbitrary equations as input and tries to rewrite using them, falling back to trying them again after applying function extensionality to also match under-applied occurences.

We could do that in `simp`. For `dunfold` and `dsimp`, this is not an option, since they must guarantee the resulting term is definitionally equal to the input term.

**Comment by Kha** [comment]
Hm, this is tricky. I'd like to avoid introducing even more tactics, but `dsimp [f]` doesn't seem like the best replacement for the current `dunfold`. Actually, isn't `dunfold` more similar to `rw`, if we taught the short-hand also to the latter tactic? That would kind of bring us back to Lean 2, where I used `rewrite` for most unfolding, and we would get (equational) refolding support for free.

In summary:
* make `unfold` apply just delta reduction (and scrap `dunfold`?)
* add the `simp` eqn short-hand to `rw`

Thoughts?

**Comment by leodemoura** [comment]
> Thoughts?

I think there a few issues:

`dsimp` and `dunfold` produce definitionally equal terms. Moreover, `dunfold [f]` unfolds all occurrences of `f`. 

`rw` produces provably equal terms, and a `rw` step rewrites a single subterm.
In principle, we could address both. For example, we would check whether the lemma used in `rw` is a "refl" lemma, if yes, then we skip the proof-term construction.

**Comment by Kha** [comment]
Ah, you are right of course. So yeah, I'm not sure any more how to fix this, or even if it needs fixing. It is not a grave inconvenience.

**Comment by avigad** [comment]
One issue is that even for ordinary definitions `f`, one might expect `simp [f]` to unfold `f`. For example, today I came across the following behavior:
``` lean
def foo (l : list ℕ) := list.map id l
def bar (l : list ℕ) := list.for l id

-- works
example : foo [1, 2, 3] = [1, 2, 3] :=
by simp [foo, list.map]

-- works
example : bar [1, 2, 3] = [1, 2, 3]:=
begin
  unfold bar,
  unfold list.for,
  unfold flip,
  simp [list.map]
end

-- doesn't work -- list.for does not unfold
example : bar [1, 2, 3] = [1, 2, 3]:=
by simp [bar, list.for, flip, list.map]

section
universe variables u v 
@[simp]
theorem list.for.equation {α : Type u} {β : Type v} (f : α → β) (l : list α) : 
  list.for l f = list.map f l := rfl
end

-- now it works
example : bar [1, 2, 3] = [1, 2, 3]:=
by simp [bar, list.for, list.map]
```
This is reasonable, but it would be convenient if there were a way to unfold ordinary definitions with `simp`. Would it be crazy to generate an `f._equation` even for ordinary instances of `def f := ...`?

**Comment by avigad** [comment]
Actually, now I am confused as to why `simp [foo]` unfold `foo`, but `simp [list.for]` doesn't unfold `list.for`.

**Comment by leodemoura** [comment]
> Would it be crazy to generate an f._equation even for ordinary instances of def f := ...?

We already generate equations for regular definitions. However, the equation is based on the signature of the function. For example:
```lean
def id (x : A) := A
```
produces the lemma
 ```lean
forall x, id x = x
```
instead of 
```lean
id = fun x, x
```

**Comment by leodemoura** [comment]
BTW, we have many tracing options for the simplifier. Examples:
```lean
set_option trace.simplify true
set_option trace.simplify.failure true
set_option trace.debug.simplify.try_rewrite true
```
We can easily find all of them using auto completion.

**Comment by leodemoura** [comment]
> Actually, now I am confused as to why simp [foo] unfold foo, but simp [list.for] doesn't unfold list.for.

The equational lemma for `list.for` is of the form. 
```lean
∀ {α : Type u} {β : Type v}, @list.for α β = ...
```
In your example, we have the term `@list.for ℕ ℕ [1, 2, 3] (λ (a : ℕ), a)`, but the simplifer does not consider the prefix `@list.for ℕ ℕ` when using the lemma above. 

I will try to consider the prefixes without sacrificing performance.

**Comment by avigad** [comment]
Thanks, that is helpful. Setting `trace.simplify` to `true` I now see `foo.equations._eqn_1` and `list.for.equations._eqn_1`.

It is counterintuitive that `list.for` is not expanded. If matching prefixes is too expensive, I guess a workaround would be to write the definition of `list.for` in expanded form. We could try to train users to do that when they want that behavior. But I agree that it would be nicer to have it happen automatically.
